minet/bluesky/http_client.py

```diff
@@ -101,30 +110,136 @@ def request(
             headers=headers,
         )
 
+        if response.status < 200 or response.status >= 300:
+            data = response.json()
+            if "error" in data:
+                e = data["error"]
+                if response.status == 400:
+                    if e == "UpstreamFailure":
+                        raise BlueskyUpstreamFailureError(
+                            "Bluesky is currently experiencing upstream issues."
+                        )
+                    elif e in (
+                        "BadRequest",
+                        "InvalidRequest",
+                        "must pass non-empty search query",
+                    ):
+                        raise BlueskyBadRequestError(data["message"])
+                raise BlueskyError(f"{e}: {data['message']}")
+            raise BlueskyError(f"HTTP {response.status}")
+
         remaining = int(response.headers["RateLimit-Remaining"])
 
         if remaining <= 0:
             self.rate_limit_reset = int(response.headers["RateLimit-Reset"])
 
         return response
 
-    def search_posts(self, query: str) -> Iterator[BlueskyPost]:
+    def search_posts(
+        self,
+        query: str,
+        lang: Optional[str] = None,
+        since: Optional[str] = None,
+        until: Optional[str] = None,
+        mentions: Optional[str] = None,
+        author: Optional[str] = None,
+        domain: Optional[str] = None,
+        url: Optional[str] = None,
+    ) -> Iterator[BlueskyPost]:
         cursor = None
-
-        while True:
-            url = self.urls.search_posts(query, cursor=cursor)
-
-            response = self.request(url)
+        oldest_post = None
+        oldest_post_uris: set[str] = set()
+        oldest_post_time_published = None
+        time_range_until: str = ""
+
+        # Search with time seems to work with millisecond precision
+        time_overlap_delta = timedelta(milliseconds=1)
+
+        # to avoid infinite loop when we find the final post,
+        # as it will always appears in the next request because of the time overlap for paging with time range
+        found_the_last_post: bool = False
+
+        while not found_the_last_post:
+
+            request_url = self.urls.search_posts(
+                query + time_range_until,
+                cursor=cursor,
+                lang=lang,
+                since=since,
+                until=until,
+                mentions=mentions,
+                author=author,
+                domain=domain,
+                url=url,
+            )
+
+            response = self.request(request_url)
             data = response.json()
 
+            if "posts" not in data or len(data["posts"]) == 0:
+                break
+            elif len(data["posts"]) == 1:
+                found_the_last_post = True
+
             for post in data["posts"]:
+                if post["uri"] in oldest_post_uris:
```

Comment:
Maybe add a comment explaining that we are relying on the uri as unique id because cid etc. cannot be trusted to be unique.